### PR TITLE
Sync android-sdk with js-sdk: Implement updates to v0.5.5

### PR DESCRIPTION
**Pull request by daeyounglnc**
#### What this PR does / why we need it?
1. Fixed GC-related TCs
Fixed TCs where garbageCollect(..) was called but GC was not actually performed

2. Converted the changes from [v0.5.5](https://github.com/yorkie-team/yorkie-js-sdk/releases/tag/v0.5.5) to the Android implementation.
- yorkie-team/yorkie-js-sdk#924 was implemented in https://github.com/yorkie-team/yorkie-android-sdk/pull/214
- [x] yorkie-team/yorkie-js-sdk#923
- [x] yorkie-team/yorkie-js-sdk#911

#### What are the relevant tickets?
-  https://github.com/yorkie-team/yorkie-android-sdk/pull/214

Fixes #

### Checklist

- [ ] Added relevant tests or not required
- [ ] Didn't break anything

## Summary by CodeRabbit

- **New Features**
  - Enhanced client configuration for improved network connectivity.
  - Introduced new error codes for refined authentication feedback.
  - Added a new message type `ArraySet` to the operation protocol.
  - Added a new test case for handling authentication errors.
  - Added a new method for publishing snapshot events with updated documents.

- **Improvements**
  - Streamlined protocol definitions and document synchronization.
  - Optimized error handling and dispatcher management for greater stability.
  - Updated document status management for consistency.
  - Improved clarity in test assertions and method calls.
  - Enhanced authentication handling within the client.

- **Tests & Chores**
  - Updated test coverage for snapshot events and edge cases.
  - Bumped project version to 0.5.5 with various minor refinements.

### CODE REVIEW

File: yorkie/src/main/kotlin/dev/yorkie/core/Interceptors.kt [hunk 1]
```diff
@@ -60,7 +64,11 @@ internal fun Client.Options.authInterceptor(): Interceptor? {
                 if (apiKey != null) {
                     put("x-api-key", listOf(apiKey))
                 }
-                if (token != null) {
+                fetchAuthToken?.let { tokenFetcher ->
+                    val token = runBlocking {
+                        tokenFetcher.invoke(shouldRefreshTokenProvider())
+                    }
+                    refreshTokenCallback?.invoke()
                     put("authorization", listOf(token))
```
[the remark this answers is not in the document]
Comment (daeyounglnc) [reply yorkie/src/main/kotlin/dev/yorkie/core/Interceptors.kt:72]: We discussed this topic and concluded that it is unlikely to cause major issues in production. If any problems arise in the future, I will consider improvements.

File: yorkie/src/test/kotlin/dev/yorkie/core/ClientTest.kt [hunk 2]
```diff
@@ -258,6 +275,23 @@ class ClientTest {
...
+        target.activateAsync().await()
+        target.attachAsync(success).await()
+        assertEquals(testAuthToken, runBlocking { target.authToken(false) })
+        assertFalse(target.shouldRefreshToken)
+        assertTrue(target.detachAsync(success).await().isSuccess)
+
+        val failing = Document(Key(AUTH_ERROR_DOCUMENT_KEY))
+        val await = target.attachAsync(failing).await()
+        assertTrue(await.isFailure)
+        assertTrue(target.shouldRefreshToken)
```
Comment (daeyounglnc) [inline yorkie/src/test/kotlin/dev/yorkie/core/ClientTest.kt:290]: (This is something I mentioned earlier) 
The test code for this part is currently failing.
I haven't found a way to properly mock the server to return that specific error.
I'd like to discuss with Musso how we can handle testing for this case.

Comment (musooff) [reply yorkie/src/test/kotlin/dev/yorkie/core/ClientTest.kt:290]: `errorCode` within `handleConnectException` is not handling properly. it is sending empty errorCode and hence shouldRefreshToken is false

Comment (daeyounglnc) [reply yorkie/src/test/kotlin/dev/yorkie/core/ClientTest.kt:290]: I briefly discussed this with you few weeks ago-though I’m not sure if you remembers :< -
I was supposed to write a test case that receives a `Code.UNAUTHENTICATED` response, but I can't do so.
If you look at the `attachDocument()` function in `MockYorkieService.kt`, the error is passed through the `ConnectException.code` field—but that was just a temporary workaround I implemented for testing purposes.
When I looked into the actual implementation in the JS SDK, it turns out the error should be passed through the `ErrorInfo` inside `ConnectException`.
However, I couldn’t figure out how to do that at the time, so I ended up creating a PR with the test code still failing.
This means it’s a part that will need to be fixed once I find the proper way to handle it.
I noticed that the JS-side test code sets up a webhook server, but since this seems more appropriate to handle via unit tests, I’ve chosen not to follow that approach.

**Comment by daeyounglnc** [comment]
> `DocumentBenchmark` test is failing. Latest updates are not applied to the file

I resolved compile error : [7180114](https://github.com/yorkie-team/yorkie-android-sdk/pull/217/commits/7180114ba6fe90d8a3fc9ddd3e184b9b0a445fc5)

**Review by musooff — APPROVED** [review]
